File: src/data/tutorials/010-minimum-window-substring.js
/**
 * Tutorial content for #010 — Minimum Window Substring
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given two strings s and t, find the smallest window (contiguous substring) in s that contains all the characters of t (including duplicates). If no such window exists, return an empty string.`,
    example: `s = "ADOBECODEBANC", t = "ABC"\n→ "BANC" contains A, B, C and has length 4\n→ "ADOBEC" also works but is longer\n✅ Answer: "BANC"`,
    keyInsight: `Use a sliding window: expand right to collect characters of t, then shrink from the left as long as the window still contains all of t. Track the minimum valid window found.`,
  },

  approaches: {
    'Brute Force': {
      intuition: `Try every possible substring of s. For each one, check if it contains all characters of t. Track the shortest valid one. Very straightforward but extremely slow because there are O(n²) substrings and each check takes O(n) time.`,
      steps: [
        `Build a frequency map for t.`,
        `Try every pair (i, j) with two nested loops.`,
        `For each substring s[i..j], check if it contains all characters of t with correct frequencies.`,
        `If valid and shorter than current best, update the answer.`,
        `Return the best substring found.`,
      ],
      example: `s = "ADOBECODEBANC", t = "ABC"\n\nCheck "A" → missing B,C ✗\nCheck "AD" → missing B,C ✗\n...\nCheck "ADOBEC" → has A,B,C ✅ len=6\n...\nCheck "BANC" → has A,B,C ✅ len=4 ← better\n...\n✅ Answer: "BANC"`,
      keyInsight: `O(n² × m) time where m = |t|. Completely impractical for long strings. Good only for understanding what "valid window" means before learning the sliding window approach.`,
    },

    'Sliding Window with Frequency Maps': {
      intuition: `Use two frequency maps: one for what t needs (required), one for what the current window has (window). Expand right, adding characters. Once the window satisfies all requirements, try to shrink from the left to find the minimum size, then expand again. The "formed" counter tracks how many distinct characters in t are currently satisfied.`,
      steps: [
        `Build requiredMap (freq of each char in t). Set required = number of unique chars in t.`,
        `Initialize left=0, formed=0, windowMap={}, best=(infinity, 0, 0).`,
        `Expand right from 0 to n-1: add s[right] to windowMap.`,
        `  If windowMap[s[right]] == requiredMap[s[right]]: formed++.`,
        `While formed == required (window is valid): try to update the best answer.`,
        `  Remove s[left] from windowMap.`,
        `  If windowMap[s[left]] < requiredMap[s[left]]: formed-- (window no longer valid).`,
        `  left++.`,
        `Return the best substring (or "" if none found).`,
      ],
      example: `s = "ADOBECODEBANC", t = "ABC"\nrequired = {A:1, B:1, C:1}, need 3 unique chars satisfied\n\nr=0(A): window={A:1}, formed=1\nr=1(D): window={A:1,D:1}\nr=2(O): window+O\nr=3(B): window={...,B:1}, formed=2\nr=4(E): window+E\nr=5(C): window={...,C:1}, formed=3 ✅ valid!\n  best="ADOBEC"(len=6), remove A(left=0)→window{A:0},formed=2,left=1\nr=6..r=9(B): window{B:2}, formed stays 2\nr=9(A): formed=3 ✅ → best="DOBECODEBA"? shrink...\n  remove D,O,B → still valid → shrink to "BANC" len=4 ✅\n✅ Answer: "BANC"`,
      keyInsight: `O(n + m) time, O(m) space. The window expands and shrinks linearly — each character is added and removed at most once. The "formed" counter avoids scanning the whole frequency map each step.`,
    },

    'Optimized with Array': {
      intuition: `Same sliding window logic, but replace the HashMap with a fixed-size integer array of length 128 (covering all ASCII characters). Array lookups are O(1) with better cache locality than a hash map. The validity check uses the same "formed" counter trick — no need to scan the full array each step.`,
      steps: [
        `Build an int[128] need[] from t (increment need[c] for each char c in t). Count distinct needed chars as required.`,
        `Initialize left=0, formed=0, bestLen=Integer.MAX_VALUE, bestLeft=0.`,
        `Expand right from 0 to n-1: decrement need[s[right]]. If need[s[right]] == 0, formed++.`,
        `While formed == required: if (right-left+1) < bestLen, update bestLen and bestLeft.`,
        `  Increment need[s[left]]. If need[s[left]] > 0, formed--. left++.`,
        `Return bestLen == MAX_VALUE ? "" : s.substring(bestLeft, bestLeft+bestLen).`,
      ],
      example: `s = "ADOBECODEBANC", t = "ABC"\nneed['A']=1, need['B']=1, need['C']=1, required=3\n\nr=0(A): need[A]=0, formed=1\nr=3(B): need[B]=0, formed=2\nr=5(C): need[C]=0, formed=3 ✅\n  bestLen=6 (ADOBEC), need[A]++ → need[A]=1>0 → formed=2, left=1\n...(slide until BANC)...\n  bestLen=4 ✅\n✅ Answer: "BANC"`,
      keyInsight: `O(n + m) time, O(1) space (array of fixed size 128). Identical asymptotic complexity to the HashMap version but faster in practice due to array cache performance — the preferred implementation in competitive settings.`,
    },
  },
}

File: public/cpp/010-minimum-window-substring.cpp
/*
 * LeetCode Problem #76: Minimum Window Substring
 * Difficulty: Hard
 * Link: https://leetcode.com/problems/minimum-window-substring/
 * 
 * Problem Statement:
 * Given two strings s and t of lengths m and n respectively, return the minimum window
 * substring of s such that every character in t (including duplicates) is included in
 * the window. If there is no such substring, return the empty string "".
 */

#include <string>
#include <unordered_map>
#include <climits>
#include <iostream>
using namespace std;

class Solution {
public:
    // ==================== APPROACH 1: Brute Force ====================
    /*
     * Algorithm:
     * - Check all possible substrings of s
     * - For each substring, check if it contains all characters of t
     * - Track minimum length substring
     * 
     * Time Complexity: O(n² * m) where n = s.length(), m = t.length()
     * Space Complexity: O(m)
     * 
     * When to use: Never (too slow)
     */
    bool containsAll(const string& s, int start, int end, unordered_map<char, int>& tCount) {
        unordered_map<char, int> windowCount;
        for (int i = start; i <= end; i++) {
            windowCount[s[i]]++;
        }
        
        for (auto& p : tCount) {
            if (windowCount[p.first] < p.second) {
                return false;
            }
        }
        return true;
    }
    
    string minWindow_BruteForce(string s, string t) {
        if (s.empty() || t.empty()) return "";
        
        unordered_map<char, int> tCount;
        for (char c : t) tCount[c]++;
        
        int minLen = INT_MAX;
        int minStart = 0;
        
        for (int i = 0; i < s.length(); i++) {
            for (int j = i; j < s.length(); j++) {
                if (containsAll(s, i, j, tCount)) {
                    if (j - i + 1 < minLen) {
                        minLen = j - i + 1;
                        minStart = i;
                    }
                    break;  // No need to extend further
                }
            }
        }
        
        return minLen == INT_MAX ? "" : s.substr(minStart, minLen);
    }
    
    // ==================== APPROACH 2: Sliding Window (Optimal) ====================
    /*
     * Algorithm:
     * 1. Use two pointers (left, right) for sliding window
     * 2. Expand right to include characters until window is valid
     * 3. Contract left to minimize window while keeping it valid
     * 4. Track minimum window found
     * 
     * Time Complexity: O(m + n) where m = s.length(), n = t.length()
     * Space Complexity: O(m + n) for hash maps
     * 
     * When to use: This is the OPTIMAL solution
     * 
     * Key Insight:
     * - Use hash map to track required characters and their counts
     * - Use 'formed' counter to track when window is valid
     * - Expand right until valid, then contract left to minimize
     * - Only check validity when adding/removing required characters
     */
    string minWindow(string s, string t) {
        if (s.empty() || t.empty()) return "";
        
        // Count characters in t
        unordered_map<char, int> tCount;
        for (char c : t) {
            tCount[c]++;
        }
        
        int required = tCount.size();  // Number of unique chars in t
        int formed = 0;  // Number of unique chars in window with desired frequency
        
        unordered_map<char, int> windowCount;
        int left = 0, right = 0;
        
        // Result: [window length, left, right]
        int minLen = INT_MAX;
        int minLeft = 0, minRight = 0;
        
        while (right < s.length()) {
            // Expand window by adding character at right
            char c = s[right];
            windowCount[c]++;
            
            // Check if frequency of current char matches desired frequency
            if (tCount.count(c) && windowCount[c] == tCount[c]) {
                formed++;
            }
            
            // Contract window from left while it's valid
            while (left <= right && formed == required) {
                // Update result if current window is smaller
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minLeft = left;
                    minRight = right;
                }
                
                // Remove character at left
                char leftChar = s[left];
                windowCount[leftChar]--;
                
                if (tCount.count(leftChar) && windowCount[leftChar] < tCount[leftChar]) {
                    formed--;
                }
                
                left++;
            }
            
            right++;
        }
        
        return minLen == INT_MAX ? "" : s.substr(minLeft, minLen);
    }
    
    // ==================== APPROACH 3: Optimized with Array ====================
    /*
     * Use array instead of hash map for ASCII characters
     * 
     * Time Complexity: O(m + n)
     * Space Complexity: O(1) - fixed size arrays
     */
    string minWindow_Array(string s, string t) {
        if (s.empty() || t.empty()) return "";
        
        vector<int> tCount(128, 0);
        int required = 0;
        
        for (char c : t) {
            if (tCount[c] == 0) required++;
            tCount[c]++;
        }
        
        vector<int> windowCount(128, 0);
        int formed = 0;
        int left = 0, right = 0;
        int minLen = INT_MAX;
        int minLeft = 0;
        
        while (right < s.length()) {
            char c = s[right];
            windowCount[c]++;
            
            if (tCount[c] > 0 && windowCount[c] == tCount[c]) {
                formed++;
            }
            
            while (left <= right && formed == required) {
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minLeft = left;
                }
                
                char leftChar = s[left];
                windowCount[leftChar]--;
                
                if (tCount[leftChar] > 0 && windowCount[leftChar] < tCount[leftChar]) {
                    formed--;
                }
                
                left++;
            }
            
            right++;
        }
        
        return minLen == INT_MAX ? "" : s.substr(minLeft, minLen);
    }
};

// ==================== TEST CASES ====================
void runTests() {
    Solution sol;
    
    // Test Case 1: Standard case
    cout << "Test 1: \"" << sol.minWindow("ADOBECODEBANC", "ABC") << "\"" << endl;
    // Expected: "BANC"
    
    // Test Case 2: No valid window
    cout << "Test 2: \"" << sol.minWindow("a", "aa") << "\"" << endl;
    // Expected: ""
    
    // Test Case 3: Entire string is minimum
    cout << "Test 3: \"" << sol.minWindow("a", "a") << "\"" << endl;
    // Expected: "a"
    
    // Test Case 4: Multiple valid windows
    cout << "Test 4: \"" << sol.minWindow("ADOBECODEBANC", "AABC") << "\"" << endl;
    // Expected: "ADOBEC" or "BANC" depending on implementation
    
    // Test Case 5: t longer than s
    cout << "Test 5: \"" << sol.minWindow("ab", "abc") << "\"" << endl;
    // Expected: ""
}

int main() {
    runTests();
    return 0;
}

/*
 * COMPARISON OF APPROACHES:
 * 
 * 1. Brute Force:
 *    Time: O(n² * m), Space: O(m)
 *    Not acceptable for interviews
 * 
 * 2. Sliding Window (RECOMMENDED):
 *    Time: O(m + n), Space: O(m + n)
 *    Optimal solution, most intuitive
 * 
 * 3. Array Optimization:
 *    Time: O(m + n), Space: O(1)
 *    Best for ASCII strings
 * 
 * INTERVIEW TIPS:
 * - Explain sliding window technique clearly
 * - Use 'formed' counter to track validity efficiently
 * - Discuss why we contract from left after expanding right
 * - Handle edge cases: empty strings, no valid window
 * - Mention that we only update result when contracting
 * 
 * KEY INSIGHTS:
 * - Sliding window pattern for substring problems
 * - Expand right to make window valid
 * - Contract left to minimize while keeping valid
 * - Use 'formed' counter instead of checking all chars each time
 * - Only care about characters that are in t
 * 
 * ALGORITHM WALKTHROUGH for s="ADOBECODEBANC", t="ABC":
 * 
 * Initial: left=0, right=0, formed=0, required=3
 * 
 * Expand phase:
 * right=0: A, formed=1 (A matched)
 * right=1: D, formed=1
 * right=2: O, formed=1
 * right=3: B, formed=2 (B matched)
 * right=4: E, formed=2
 * right=5: C, formed=3 (C matched) -> Valid!
 * 
 * Contract phase:
 * Window "ADOBEC" (len=6), update min
 * left=1: Remove A, formed=2 -> Invalid
 * 
 * Continue expanding...
 * Eventually find "BANC" (len=4) which is minimum
 * 
 * COMMON MISTAKES:
 * - Not handling duplicate characters in t
 * - Checking validity on every iteration (inefficient)
 * - Wrong condition for contracting window
 * - Not updating result during contraction
 * - Forgetting edge cases
 * 
 * FOLLOW-UP QUESTIONS:
 * - Find all minimum windows? (Store all with min length)
 * - Case insensitive? (Convert to lowercase first)
 * - Unicode characters? (Use hash map, not array)
 * - Return indices instead of substring? (Return [minLeft, minRight])
 * 
 * RELATED PROBLEMS:
 * - Longest Substring Without Repeating Characters (LeetCode #3)
 * - Longest Substring with At Most K Distinct Characters (LeetCode #340)
 * - Substring with Concatenation of All Words (LeetCode #30)
 * - Find All Anagrams in a String (LeetCode #438)
 * 
 * SLIDING WINDOW PATTERN:
 * 1. Expand right to include more elements
 * 2. Check if window is valid
 * 3. If valid, try to contract from left
 * 4. Update result when appropriate
 * 5. Repeat until right reaches end
 */

// Made with Bob
